File: solutions/1160-letter-tile-possibilities/solution.cpp
class Solution {
public:
    void solve(string tiles,unordered_set<string>&s,string word){
        s.insert(word);
        if(tiles==""){
            return;
        }
        // if(s.count(word)){
        //     return;
        // }
        string tiles_copy = tiles;
        for(int i=0;i<tiles.size();i++){
            char c = tiles[i];
            string newWord = word + c;
            tiles = tiles.erase(i,1);
            solve(tiles,s,newWord);
            tiles = tiles_copy;
        }
    }
    int numTilePossibilities(string tiles) {
        unordered_set<string> s;
        solve(tiles,s,"");
        return s.size()-1;
    }
};

File: solutions/1160-letter-tile-possibilities/solution.py
class Solution:
    def _solve(self,tiles:str,s:set,word:str) -> None:
        s.add(word)
        tiles_copy = tiles
        for i in range(0,len(tiles_copy)):
            tiles = tiles[:i] + tiles[i+1:]
            self._solve(tiles,s,word+tiles_copy[i])
            tiles = tiles_copy

    def numTilePossibilities(self, tiles: str) -> int:
        s = set()
        self._solve(tiles,s,"")
        return len(s)-1
    


        
